File: app/app/src/main/java/me/williamhester/reddit/ui/viewholders/SubmissionSelfTextViewHolder.java
package me.williamhester.reddit.ui.viewholders;

import android.view.View;
import android.widget.TextView;

import me.williamhester.reddit.R;
import me.williamhester.reddit.html.HtmlParseResult;
import me.williamhester.reddit.models.Submission;
import me.williamhester.reddit.html.HtmlParser;
import me.williamhester.reddit.ui.VotableCallbacks;
import me.williamhester.reddit.ui.text.VotableMovementMethod;

import static butterknife.ButterKnife.findById;

/** A ViewHolder for a {@link Submission} that has self text. */
public class SubmissionSelfTextViewHolder extends SubmissionViewHolder {

  private final HtmlParser htmlParser;
  private final TextView selfText;
  private final VotableMovementMethod linkMovementMethod;

  public SubmissionSelfTextViewHolder(
      View itemView, HtmlParser htmlParser, VotableCallbacks clickListener) {
    super(itemView, clickListener);

    this.htmlParser = htmlParser;
    this.selfText = findById(itemView, R.id.self_text);
    this.linkMovementMethod = new VotableMovementMethod(clickListener);
    this.selfText.setMovementMethod(linkMovementMethod);
  }

  @Override
  public void setContent(Submission item) {
    super.setContent(item);

    linkMovementMethod.setVotable(item);
    HtmlParseResult result = htmlParser.parseHtml(item.getSelftextHtml());
    selfText.setText(result.getText());
  }
}

File: app/app/src/main/java/me/williamhester/reddit/ui/viewholders/SubmissionSelfTextViewHolder.kt
package me.williamhester.reddit.ui.viewholders

import android.view.View
import android.widget.TextView

import me.williamhester.reddit.R
import me.williamhester.reddit.models.Submission
import me.williamhester.reddit.html.HtmlParser
import me.williamhester.reddit.ui.VotableCallbacks
import me.williamhester.reddit.ui.text.VotableMovementMethod

/** A ViewHolder for a [Submission] that has self text.  */
class SubmissionSelfTextViewHolder(
    itemView: View,
    clickListener: VotableCallbacks,
    private val htmlParser: HtmlParser
) : SubmissionViewHolder(itemView, clickListener) {

  private val selfText: TextView = itemView.findViewById(R.id.self_text)
  private val linkMovementMethod: VotableMovementMethod = VotableMovementMethod(clickListener)

  init {
    this.selfText.movementMethod = linkMovementMethod
  }

  override fun setContent(item: Submission) {
    super.setContent(item)

    linkMovementMethod.setVotable(item)
    val result = htmlParser.parseHtml(item.selftextHtml)
    selfText.text = result.text
  }
}
